src/utils/config.py

```diff
@@ -0,0 +1,246 @@
+
+
+def _validate_path(path: str, param_nm: str, config: dict):
+    """Validate a passed path (str format) is valid.
+
+    Args:
+        path (str): The path to validate.
+        param_nm (str): The param name (for error raises).
+        item_conf (dict): The config file (for file ext).
+    """
+    _validate_str_is_path(path, param_nm)
+    try:
+        file_ext = config["filetype"]
+    except KeyError:
+        file_ext = None
+    if file_ext:
+        _validate_file_extension(path, file_ext)
+
+
+def _validate_numeric(value: Union[float, int], param_nm: str, config: dict):
+    """Valdiate a numerical value fits between a range.
+
+    Args:
+        value (Union[float, int]): The numerical value.
+        param_nm (str): The parameter name (used for error raises).
+        config (dict): The config.
+
+    Raises:
+        ValueError: Raised if a number is lesser than the minimum.
+        ValueError: Raised if a number is greater than the maximum.
+    """
+    try:
+        max = config["max"]
+    except KeyError:
+        max = None
+    try:
+        min = config["min"]
+    except KeyError:
+        min = None
+    if max:
+        if value > max:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) greater than max ({max})."
+            )
+    if min:
+        if value < min:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) less than min ({min})."
+            )
+
+
+def _nulltype_conversion(value: str) -> Union[str, None]:
+    """Account for user disparities in the form of NoneType they pass.
+
+    Args:
+        value (str): The value passed by a user.
+
+    Returns:
+        Union[str, None]: The altered value.
+    """
+    _type_defence(value, "value", str)
+    if value.lower() in ["", "null", "none"]:
+        return None
+    return value
+
+
+def _check_items(item: dict, config_item: dict, item_name: str) -> None:
+    """Check items of a config to validate them.
+
+    Args:
+        item (dict): The item(s) to check.
+    """
+    DTYPES = {
+        "int": int,
+        "float": float,
+        "str": str,
+        "bool": bool,
+        "path": str, # passed as string in config
+        "list": list
+    }
+    # recursive check on schema item
+    if not _check_has_schema(value=item):
+        for sub_item in item.keys():
+            _check_items(
+                item=item[sub_item],
+                config_item=config_item[sub_item],
+                item_name=sub_item
+            )
+        return
+    # check that both levels aren't the same
+    if item["top_level"] == item["bottom_level"]:
+        raise ValueError(
+            "Conflicting config. top_level == botom_level"
+        )
+    # do validation
+    if item["bottom_level"]:
+        config_item = {item_name: config_item}
+    dtype = DTYPES[item["dtype"]] if "list" not in item["dtype"] else list
+    if item["accept_nonetype"]:
+        dtype = (dtype, type(None))
```

Comment:
what does this do?

Reply:
This adds the NoneType to the type defence check (accept_nonteype)